runtime/moonbase/tests/xcm_testing/relay_asset.rs

```diff
@@ -0,0 +1,320 @@
+
+	ParaA::execute_with(|| {
+		let asset = currency_to_asset(parachain::CurrencyId::ForeignAsset(relay_asset_id), 100);
+		assert_ok!(PolkadotXcm::transfer_assets(
+			parachain::RuntimeOrigin::signed(PARAALICE.into()),
+			Box::new(VersionedLocation::from(chain_part)),
+			Box::new(VersionedLocation::from(beneficiary)),
+			Box::new(VersionedAssets::from(vec![asset])),
+			0,
+			WeightLimit::Limited(Weight::from_parts(40000u64, DEFAULT_PROOF_SIZE))
+		));
+	});
+
+	// Verify balances
+	assert_asset_balance(&PARAALICE, relay_asset_id, 23);
+	assert_asset_balance_para_b(&PARAALICE, relay_asset_id, 100);
+}
+
+#[test]
+fn receive_relay_asset_with_trader() {
+	MockNet::reset();
+
+	// Use helper for high units_per_second registration
+	let relay_asset_id = register_relay_asset_with_units_per_second(2_500_000_000_000);
+
+	let dest: Location = Junction::AccountKey20 {
+		network: None,
+		key: PARAALICE,
+	}
+	.into();
+	// We are sending 100 tokens from relay.
+	// Amount spent in fees is Units per second * weight / 1_000_000_000_000 (weight per second)
+	// weight is 4 since we are executing 4 instructions with a unitweightcost of 1.
+	// Units per second should be 2_500_000_000_000_000
+	// Therefore with no refund, we should receive 10 tokens less
+	// Native trader fails for this, and we use the asset trader
+	Relay::execute_with(|| {
+		assert_ok!(RelayChainPalletXcm::limited_reserve_transfer_assets(
+			relay_chain::RuntimeOrigin::signed(RELAYALICE),
+			Box::new(Parachain(1).into()),
+			Box::new(VersionedLocation::from(dest).clone()),
+			Box::new(([], 100).into()),
+			0,
+			WeightLimit::Unlimited
+		));
+	});
+
+	// Use assertion
+	assert_asset_balance(&PARAALICE, relay_asset_id, 90);
+	assert_treasury_asset_balance(relay_asset_id, 10);
+}
+
+#[test]
+fn error_when_not_paying_enough() {
+	MockNet::reset();
+
+	let relay_asset_id = register_relay_asset_with_units_per_second(2500000000000);
+
+	let dest: Location = Junction::AccountKey20 {
+		network: None,
+		key: PARAALICE,
+	}
+	.into();
+
+	// Initial state verification - should be zero
+	assert_asset_balance(&PARAALICE, relay_asset_id, 0);
+
+	// We are sending 5 tokens from relay - not enough to pay fees
+	// If we set the dest weight to be 1e7, we know the buy_execution will spend 1e7*1e6/1e12 = 10
+	// Therefore with no refund, we should receive 10 tokens less
+	Relay::execute_with(|| {
+		assert_ok!(RelayChainPalletXcm::limited_reserve_transfer_assets(
+			relay_chain::RuntimeOrigin::signed(RELAYALICE),
+			Box::new(Parachain(1).into()),
+			Box::new(VersionedLocation::from(dest).clone()),
+			Box::new(([], 5).into()),
+			0,
+			WeightLimit::Unlimited
+		));
+	});
+
+	// Use helper for assertion - amount not received as it is not paying enough
+	assert_asset_balance(&PARAALICE, relay_asset_id, 0);
+}
+
+#[test]
+fn receive_asset_with_no_sufficients_not_possible_if_non_existent_account() {
```

Comment:
Original https://github.com/moonbeam-foundation/moonbeam/blob/fb5af73509cccc5ecef95de28896679b7cef11c5/runtime/moonbase/tests/xcm_tests.rs#L2322C1-L2322C27

Reply:
This could be moved to `evm_account` file